pkg/operator/workload/workload_openshiftapiserver_v311_00_sync_test.go

```diff
@@ -219,3 +237,155 @@ func TestPreconditionFulfilled(t *testing.T) {
+				&corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: "etcd-client", Namespace: operatorclient.GlobalUserSpecifiedConfigNamespace}},
+				&appsv1.Deployment{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:       "apiserver",
+						Namespace:  "openshift-apiserver",
+						Generation: 100,
+					},
+					Status: appsv1.DeploymentStatus{
+						AvailableReplicas: 100,
+					},
+				})
+
+			operatorConfig := &operatorv1.OpenShiftAPIServer{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:       "cluster",
+					Generation: 100,
+				},
+				Spec: operatorv1.OpenShiftAPIServerSpec{
+					OperatorSpec: operatorv1.OperatorSpec{},
+				},
+				Status: operatorv1.OpenShiftAPIServerStatus{
+					OperatorStatus: operatorv1.OperatorStatus{
+						ObservedGeneration: 100,
+					},
+				},
+			}
+			apiServiceOperatorClient := operatorfake.NewSimpleClientset(operatorConfig)
+			openshiftConfigClient := configfake.NewSimpleClientset(
+				&configv1.Image{ObjectMeta: metav1.ObjectMeta{Name: "cluster"}},
+				&configv1.ClusterVersion{
+					ObjectMeta: metav1.ObjectMeta{Name: "version"},
+					Status: configv1.ClusterVersionStatus{
+						Capabilities: configv1.ClusterVersionCapabilitiesStatus{
+							EnabledCapabilities: tc.enabledCapabilities,
+							KnownCapabilities:   tc.knownCapabilities,
+						},
+					},
+				},
+			)
+
+			fakeOperatorClient := operatorv1helpers.NewFakeOperatorClient(&operatorv1.OperatorSpec{ManagementState: operatorv1.Managed}, &operatorv1.OperatorStatus{}, nil)
+			configInformers := configinformers.NewSharedInformerFactory(openshiftConfigClient, 10*time.Minute)
+
+			target := OpenShiftAPIServerWorkload{
+				kubeClient:                kubeClient,
+				operatorClient:            fakeOperatorClient,
+				operatorConfigClient:      apiServiceOperatorClient.OperatorV1(),
+				openshiftConfigClient:     openshiftConfigClient.ConfigV1(),
+				clusterVersionInformer:    configInformers.Config().V1().ClusterVersions(),
+				versionRecorder:           status.NewVersionGetter(),
+				countNodes:                fakeCountNodes,
+				ensureAtMostOnePodPerNode: func(spec *appsv1.DeploymentSpec, componentName string) error { return nil },
+			}
+
+			// register the informer
+			configInformers.Config().V1().ClusterVersions().Informer()
+
+			ctx := context.Background()
+			configInformers.Start(ctx.Done())
+			configInformers.WaitForCacheSync(ctx.Done())
+
+			if _, _, err := target.Sync(ctx, factory.NewSyncContext("TestSyncCOntext", events.NewInMemoryRecorder(""))); len(err) > 0 {
+				t.Fatal(err)
+			}
+
+			cm, err := kubeClient.CoreV1().ConfigMaps("openshift-apiserver").Get(ctx, "config", metav1.GetOptions{})
+			if err != nil {
+				t.Fatalf("Unable to get 'config' configmap: %v", err)
+			}
+
+			config := openshiftcontrolplanev1.OpenShiftAPIServerConfig{}
+			if err := json.NewDecoder(bytes.NewBuffer([]byte(cm.Data["config.yaml"]))).Decode(&config); err != nil {
```

Comment:
`oas` uses a versioned decoder. Could we do the same here ?

```
scheme := runtime.NewScheme()
utilruntime.Must(openshiftcontrolplanev1.Install(scheme))
codecs := serializer.NewCodecFactory(scheme)
obj, err := runtime.Decode(codecs.UniversalDecoder(openshiftcontrolplanev1.GroupVersion, configv1.GroupVersion), configContent)
if err != nil {
	return err
}
config := obj.(*openshiftcontrolplanev1.OpenShiftAPIServerConfig)
```


https://github.com/openshift/openshift-apiserver/blob/df3ca642f426cf2f34abeae152a9fce80b44c63e/pkg/cmd/openshift-apiserver/cmd.go#L129

Reply:
Done